addons/l10n_it_edi/tests/test_ir_mail_server.py

```diff
@@ -158,3 +180,149 @@ def test_ricevuta_consegna(self):
+                }),
+                (0, 0, {
+                    'name': 'something else price included',
+                    'quantity': 1,
+                    'price_unit': 800.40,
+                    'tax_ids': [(6, 0, [self.price_included_tax.id])]
+                }),
+                (0, 0, {
+                    'name': 'something not price included',
+                    'quantity': 1,
+                    'price_unit': 800.40,
+                    'tax_ids': [(6, 0, [self.tax_sale_a.id])]
+                }),
+            ],
+        })
+
+        # In this case, the first two lines use a price_include tax the
+        # subtotals should be 800.40 / (100 + 22.0) * 100 = 656.065564..,
+        # where 22.0 is the tax rate.
+        #
+        # Since the subtotals are rounded we actually have 656.07
+        lines = price_included_invoice.line_ids
+        price_included_lines = lines.filtered(lambda line: line.tax_ids == self.price_included_tax)
+        self.assertEqual([line.price_subtotal for line in price_included_lines], [656.07, 656.07])
+        # So the taxable a base the edi expects (for this tax) is actually 1312.14
+        price_included_tax_line = lines.filtered(lambda line: line.tax_line_id == self.price_included_tax)
+        self.assertEqual(price_included_tax_line.tax_base_amount, 1312.14)
+
+        # The tax amount of the price included tax should be:
+        #   per line: 800.40 - (800.40 / (100 + 22) * 100) = 144.33
+        #   tax amount: 144.33 * 2 = 288.66
+        self.assertEqual(price_included_tax_line.price_total, 288.66)
+
+        # Prepare the XML
+        template_values = price_included_invoice._prepare_fatturapa_export_values()
+        xml = self.env.ref('l10n_it_edi.account_invoice_it_FatturaPA_export')._render(template_values)
+        tree = etree.fromstring(xml)
+
+        # The EDI constraint is that base amount * tax rate = tax amount
+        # and since 1312.14 * 0.22 = 288.67, and since 288.67 != 288.66
+        # our base amount should not be equal to 1312.14
+        # Our edi should contain a base_amount value such that:
+        #   base_amount * 0.22 = 288.66
+        tax_bases = [float(base.text) for base in tree[1].xpath('.//ImponibileImporto')]
+        tax_bases.sort()
+        # 1312.09 * 0.22 = 288.65979999... = 288.66 for our purposes. 800.40 is the base of the price excluded tax
+        self.assertEqual(tax_bases, [800.40, 1312.09])
+
+        # A secondary edi constraint is that the sum(price_subtotal) + rounding_amount = base_amount
+        xml_subtotals = [float(sub.text) for sub in tree[1].xpath('.//PrezzoUnitario')]
+        xml_subtotals.sort()
+        self.assertEqual(xml_subtotals, [656.07, 656.07, 800.4])
+
+        rounding_amount = float(tree[1].xpath('.//Arrotondamento')[0].text)
+        self.assertEqual(
+            tools.float_round(656.07 + 656.07 + rounding_amount, precision_rounding=self.company_data['currency'].rounding),
```

Comment:
@dbkosky Why would not we check the entire file at once?  (check assertXmlTreeEqual also for MX and PE)

Reply:
@dbkosky But the rest of the xml does matter when we send it to the government.  We can also send a standard XML as the standard test case and then for the private xml, we change that xml with an xpath (and would be similar to how you put it here)